### ISSUE TITLE
BlockSparseMatrixEZ is not derived from BlockMatrixBase

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#11242]
Unlike all of the other `Block*Matrix` classes, `BlockSparseMatrixEZ` is going it alone and implements everything itself. That is unnecessary and probably duplicates a bunch of code. We should just derive it from `BlockMatrixBase` in the same way as, for example, `BlockSparseMatrix`.

**Comment by FelixKimmerle** [comment]
Hi, I would like to start working on this issue.
If someone else is already working on it, please let me know.

My plan:
- Identify which methods and members can be reused from `BlockMatrixBase` (mirroring `BlockSparseMatrix`)
- Make `BlockSparseMatrixEZ` derive from `BlockMatrixBase`
- Remove duplicated code while keeping the API and behavior unchanged
- Run and extend tests as needed
- Check comments and documentation for outdated or redundant references to `BlockSparseMatrixEZ`

One question before I start: is there any historical/design reason why `BlockSparseMatrixEZ` does not currently derive from `BlockMatrixBase` (e.g., past constraints or edge cases) that I should be aware of?

**Comment by bangerth** [comment]
@FelixKimmerle  It would be fantastic if you'd work on this! I think your plan sounds exactly like I'd approach this too.

I don't recall why `BlockSparseMatrixEZ` is not derived from `BlackMatrixBase` -- it sounds like a historical accident. If I had to guess, the issue arose at a time when we only had two matrix classes, `SparseMatrix` and `SparseMatrixEZ`, and that when we introduced block matrices, we created separate `BlockSparseMatrix` and `BlockSparseMatrixEZ`. When the number of matrix classes started to proliferate when the PETSc and Trilinos interfaces were written, we broke `BlockMatrixBase` out of `BlockSparseMatrix` and forgot to do the same for `BlockSparseMatrixEZ`. But all of this happened 20+ years ago, and I don't think anyone still remembers without actually going through the git history...

**Comment by bangerth** [comment]
Let us know if you run into issues you'd like to discuss -- either here or on the mailing list!

**Comment by FelixKimmerle** [comment]
Hello Wolfgang (@bangerth) and others.
Quick status update + a design/scope question that came up while trying to add meaningful test coverage.

I made `BlockSparseMatrixEZ` derive from `BlockMatrixBase` and removed the obvious code duplication. Where needed for the refactor, I also aligned a few small pieces of functionality with what exists in `BlockSparseMatrix`/`SparseMatrix`. The library compiles.

However, I ran into a testing/usage problem: `BlockSparseMatrixEZ` (and `SparseMatrixEZ`) appear to be essentially unused in the current code base (no tests/examples/library code that uses them). To add coverage, I tried porting `tests/lac/block_matrices_04.cc` (Laplace) to use `BlockSparseMatrixEZ`/`SparseMatrixEZ`, since it has known output and would be a good indicator that the refactor didn’t change behavior.

That attempt exposed missing functionality needed for a “normal” workflow:

- `apply_boundary_values()` is not available for the EZ classes.
- Trying to implement that revealed further API gaps compared to the non-EZ sparse matrices (e.g., `SparseMatrixEZ` lacked `diag_element()`, which I added).
- The bigger blocker is that `SparseMatrixEZ` does not provide writable iterators, so common patterns that modify entries by using a writable iterator (which for example `apply_boundary_values()` does) can’t be ported in a straightforward way.

At this point I’m unsure what the intended use / supported feature set of the EZ matrices is. Concretely: are `SparseMatrixEZ` / `BlockSparseMatrixEZ` meant to be convenience classes with a reduced API (not expected to support the same workflows as `SparseMatrix` / `BlockSparseMatrix`), or are they intended to be almost drop-in alternatives where one can do essentially everything, just without explicitly providing a sparsity pattern?

Depending on the answer, I see two ways forward:

1. **Minimal scope (refactor only):** Keep the refactor (derive from `BlockMatrixBase`, remove duplication) and add a small test for `BlockSparseMatrixEZ` that exercises basic functionality (set/read entries, simple linalg ops), without trying to support `apply_boundary_values()`/standard assembly workflows.
2. **Basic usability support:** Extend the EZ classes enough so typical workflows (assembly + boundary value application + solve) work, and then add a real test like the Laplace one. This requires non-trivial additions (e.g. writable iterator support or an alternative implementation of `apply_boundary_values()`) and may exceed the original issue’s scope.

Could you (or others) clarify which direction matches the project’s intent? In particular: should `apply_boundary_values()` be expected to work with `SparseMatrixEZ`/`BlockSparseMatrixEZ`?

Thanks for your time, and for any thoughts on what direction you’d prefer here.
Appreciate the guidance!

**Comment by FelixKimmerle** [comment]
I've thought about this some more and believe that the best course of action is a minimal-scope refactor.

I will shortly open a pull request that implements the changes discussed above. If making `BlockSparseMatrixEZ` fully compatible with `BlockSparseMatrix` (for use in the Laplace problem) is desired, I think this should be handled in a separate issue, as it goes beyond the scope of this issue.

Please let me know whether you would prefer that I open such an issue myself. 

I will include the information I discovered while trying to make this work, in the new issue.

**Comment by masterleinad** [comment]
If `BlockSparseMatrixEZ` isn't used anywhere (implying that it might not have any good use cases), one option would also be to just deprecate it to remove it later.

**Comment by bangerth** [comment]
I agree that a minimal refactor is the way to go. We can always make the class fully functional as a follow-up, but I find incremental patches a lot easier to deal with. Thank you for working on it this way!

As for the class not being used: `SparseMatrixEZ` is one of our oldest classes. It predates many of the other, more systematic, ways we've transformed our linear algebra classes over the past 20 or more years, and it never found the same kind of use `SparseMatrix` has. I don't think that's necessarily because the class's design is bad, it's just that we tend to use standardized approaches that are the same in every program because standardization makes things easier to learn. (Another class in this vein is `ChunkSparseMatrix`, which I continue to think is conceptually right but not used in practice because it's "different".)

So it is true that we could, without much loss, also just deprecate it and get rid of. I don't really have much of an opinion about that, whereas I think it is a good incremental step to make the class use the same structure as all other classes of the same type.

**Comment by bangerth** [comment]
@masterleinad Separately: We do have a bunch of tests for `SparseMatrixEZ`, just not for `BlockSparseMatrixEZ`. I think if we keep the former, we should also keep the latter. (And if we decided to deprecate the latter, we should also deprecate the former.)

**Comment by bangerth** [comment]
Fixed by #19218.
